examples/authentication/authenticate_in_web_application.py

```diff
@@ -12,60 +12,184 @@
 
 
 def main(client_secrets_path, scopes):
-    flow = InstalledAppFlow.from_client_secrets_file(
-        client_secrets_path, scopes=scopes
+    """The main method, starts a basic server and initializes an auth request.
+
+    Args:
+        client_secrets_path: a path to where the client secrete JSON file is
+            located on the machine running this example.
+        scopes: a list of API scopes to include in the auth request, see:
+            https://developers.google.com/identity/protocols/oauth2/scopes
+    """
+    flow = Flow.from_client_secrets_file(client_secrets_path, scopes=scopes)
+    flow.redirect_uri = _REDIRECT_URI
+
+    # Create an anti-forgery state token as described here:
+    # https://developers.google.com/identity/protocols/OpenIDConnect#createxsrftoken
+    passthrough_val = hashlib.sha256(os.urandom(1024)).hexdigest()
+
+    authorization_url, state = flow.authorization_url(
+        access_type='offline',
+        state=passthrough_val,
+        include_granted_scopes='true'
+    )
+
+    # Prints the authorization URL so you can paste into your browser. In a
+    # typical web application you would redirect the user to this URL, and they
+    # would be redirected back to "redirect_url" provided earlier after
+    # granting permission.
+    print(authorization_url)
+
+    try:
+        # Retrieves an authorization code by opening a socket to receive the
+        # redirect request and parsing the query parameters set in the URL.
+        code = _get_authorization_code(passthrough_val)
+    except ValueError as error:
+        print(error)
+        sys.exit(1)
+
+    # Pass the code back into the OAuth module to get a refresh token.
+    flow.fetch_token(code=code)
+    refresh_token = flow.credentials.refresh_token
+
+    print(f"\nYour refresh token is: {refresh_token}\n")
+    print(
+        "Add your refresh token to your client library configuration as "
+        "described here: "
+        "https://developers.google.com/google-ads/api/docs/client-libs/python/configuration"
     )
 
-    flow.run_local_server()
 
-    print(f"Access token: {flow.credentials.token}")
-    print(f"Refresh token: {flow.credentials.refresh_token}")
+def _get_authorization_code(passthrough_val):
+    """Opens a socket to handle a single HTTP request containing auth tokens.
+
+    Args:
+        passthrough_val: an anti-forgery token used to verify the request
+            received by the socket.
+
+    Returns:
+        a str access token from the Google Auth service.
+    """
+    # Open a socket at localhost:8080 and listen for a request
+    sock = socket.socket()
+    sock.bind(('localhost', 8080))
+    sock.listen(1)
+    connection, address = sock.accept()
+    data = connection.recv(1024)
+    # Parse the raw request to retrieve the URL query parameters.
+    params = _parse_raw_query_params(data)
+
+    try:
+        if not params.get("code"):
+            message = "Failed to retrieve authorization code."
```

Comment:
Are there any additional debug resources we can point to for this in the message?

Reply:
Not in addition to what's already linked, but according to [this page](https://developers.google.com/identity/protocols/oauth2/web-server#handlingresponse) if the request doesn't contain an auth token, it will contain an error message, so I'll include that in this string.